### PR TITLE
Feature/config validation

### PR DESCRIPTION
**Pull request by JulioV**
Review

### CODE REVIEW

File: tools/config.schema.yaml [hunk 1]
```diff
@@ -0,0 +1,855 @@
+$schema: "http://json-schema.org/draft-07/schema#"
+description: RAPIDS configuration schema
+required:
```
Comment (JulioV) [inline tools/config.schema.yaml:3]: Add all existent keys so they are mandatory

File: tools/config.schema.yaml [hunk 2]
```diff
@@ -0,0 +1,855 @@
...
+properties:
+  DATABASE_GROUP:
+    type: string
+
+  TIMEZONE:
+    type: string
+  
+  CREATE_PARTICIPANT_FILES:
+    type: object
+    required: [SOURCE]
```
Comment (JulioV) [inline tools/config.schema.yaml:95]:
```suggestion
    required: [SOURCE, PHONE_SECTION, FITBIT_SECTION]
```

File: tools/config.schema.yaml [hunk 3]
```diff
@@ -0,0 +1,855 @@
...
+    properties:
+      TABLE:
+        type: object
+        required: [ANDROID, IOS]
+        properties:
+          ANDROID: 
+            type: string
+          IOS: 
+            type: string
+      EPISODE_THRESHOLD_BETWEENROWS:
```
Comment (JulioV) [inline tools/config.schema.yaml:193]:
```suggestion
      EPISODE_THRESHOLD_BETWEEN_ROWS:
```

File: tools/config.schema.yaml [hunk 4]
```diff
@@ -0,0 +1,855 @@
...
+        type: integer
+        maximum: 5
```
Comment (JulioV) [inline tools/config.schema.yaml:195]:
```suggestion
        maximum: 1440
```

File: tools/config.schema.yaml [hunk 5]
```diff
@@ -0,0 +1,855 @@
...
+        uniqueItems: True
+        items:
+          type: string
+          enum: ["withinsegments", "acrosssegments", "acrossdataset"]
+  
+  APPLICATION_CATEGORIES_FEATURE:
+        type: object
+        required: [CATALOGUE_SOURCE, CATALOGUE_FILE, UPDATE_CATALOGUE_FILE, SCRAPE_MISSING_CATEGORIES]
+        properties:
+          CATALOGUE_SOURCE:
```
Comment (JulioV) [inline tools/config.schema.yaml:53]: this one has to be an enum FILE or GOOGLE

File: tools/config.schema.yaml [hunk 6]
```diff
@@ -0,0 +1,855 @@
...
+          RAPIDS: 
+            allOf:
+              - $ref: "#/definitions/PROVIDER"
+              - properties:
+                  SINGLE_CATEGORIES: 
+                    type: array
+                    uniqueItems: True
+                    items:
+                      type: string
+                      enum: ["all", "email"]
```
Comment (JulioV) [inline tools/config.schema.yaml:271]: This one shouldn't be an enum, the name of the categories are arbitrary

File: tools/config.schema.yaml [hunk 7]
```diff
@@ -0,0 +1,855 @@
...
+                  MULTIPLE_CATEGORIES:
+                    type: object
+                    required: [social, entertainment]
```
Comment (JulioV) [inline tools/config.schema.yaml:274]: same as before, there are no required values

File: tools/config.schema.yaml [hunk 8]
```diff
@@ -0,0 +1,855 @@
...
+                    properties:
+                      social:
+                        type: array
+                        uniqueItems: True
+                        items:
+                          type: string
+                          enum: ["socialnetworks", "socialmediatools"]
```
Comment (JulioV) [inline tools/config.schema.yaml:281]: idem, not an enum MULTIPLE_CATEGORIES should be an object of string arrays

File: tools/config.schema.yaml [hunk 9]
```diff
@@ -0,0 +1,855 @@
...
+                    type: ["null", array]
+                    uniqueItems: True
+                    items:
+                      type: string
+                  EXCLUDED_APPS: 
+                    type: array
+                    uniqueItems: true
+                    items:
+                      type: string
+                      enum: ["com.fitbit.FitbitMobile", "com.aware.plugin.upmc.cancer"]
```
Comment (JulioV) [inline tools/config.schema.yaml:298]: not an enum

File: tools/config.schema.yaml [hunk 10]
```diff
@@ -0,0 +1,855 @@
...
+        
+  PHONE_BATTERY:
+    type: object
+    required: [TABLE, PROVIDERS, EPISODE_THRESHOLD_BETWEEN_ROWS]
+    properties:
+      TABLE:
+        type: string
+      EPISODE_THRESHOLD_BETWEEN_ROWS:
+        type: integer
+        maximum: 30
```
Comment (JulioV) [inline tools/config.schema.yaml:338]:
```suggestion
        maximum: 1440
```

File: tools/config.schema.yaml [hunk 11]
```diff
@@ -0,0 +1,855 @@
...
+            allOf:
+              - $ref: "#/definitions/PROVIDER"
+              - properties:
+                  FEATURES:
+                    type: object
+                    required: [ALL, OWN, OTHERS]
+                    properties:
+                      ALL: 
+                        allOf:
+                          - $ref: "#/definitions/DORYAB_BLUETOOTH_FEATURE"
```
Comment (JulioV) [inline tools/config.schema.yaml:383]: Nice reuse!

File: tools/config.schema.yaml [hunk 12]
```diff
@@ -0,0 +1,855 @@
...
+                    items:
+                      type: string
+                      enum: ["minutessilence", "minutesnoise", "minutesvoice", "minutesunknown","sumconversationduration","avgconversationduration",
+    "sdconversationduration","minconversationduration","maxconversationduration","timefirstconversation","timelastconversation","noisesumenergy",
+    "noiseavgenergy","noisesdenergy","noiseminenergy","noisemaxenergy","voicesumenergy",
+    "voiceavgenergy","voicesdenergy","voiceminenergy","voicemaxenergy","silencesensedfraction","noisesensedfraction",
+    "voicesensedfraction","unknownsensedfraction","silenceexpectedfraction","noiseexpectedfraction","voiceexpectedfraction",
+    "unknownexpectedfraction","countconversation"]
+                  RECORDING_MINUTES: 
+                    type: integer
```
Comment (JulioV) [inline tools/config.schema.yaml:469]: min 1, max 1440

File: tools/config.schema.yaml [hunk 13]
```diff
@@ -0,0 +1,855 @@
...
+                  PAUSED_MINUTES:
+                    type: integer
```
Comment (JulioV) [inline tools/config.schema.yaml:471]: min 1, max 1440

File: tools/config.schema.yaml [hunk 14]
```diff
@@ -0,0 +1,855 @@
...
+          $ref: "#/definitions/PROVIDER"
+  
+  PHONE_DATA_YIELD:
+    type: object
+    required: [SENSORS, PROVIDERS]
+    properties:
+      SENSORS:
+        type: ["null", array]
+        items:
+          type: string
```
Comment (JulioV) [inline tools/config.schema.yaml:482]: SENSORS should be an enum with possible values equal to any PHONE_* sensor except for PHONE_DATA_YIELD itself

File: tools/config.schema.yaml [hunk 15]
```diff
@@ -0,0 +1,855 @@
...
+      FUSED_RESAMPLED_TIME_SINCE_VALID_LOCATION:
+        type: integer
+        exclusiveMinimum: 0
+      PROVIDERS:
+        type: ["null", object]
+        properties:
+          DORYAB:
+            allOf:
+                - $ref: "#/definitions/PROVIDER"
+                - properties:
```
Comment (JulioV) [inline tools/config.schema.yaml:555]: the rest of the config parameters are missing like DBSCAN_EPS

File: tools/config.schema.yaml [hunk 16]
```diff
@@ -0,0 +1,855 @@
...
+                    FEATURES: 
+                      type: array
+                      uniqueItems: True
+                      items:
+                        type: string
+                        enum: [locationvariance,loglocationvariance,totaldistance,averagespeed,varspeed,circadianmovement,numberofsignificantplaces,numberlocationtransitions,radiusgyration,timeattop1location,timeattop2location,timeattop3location,movingtostaticratio,outlierstimepercent,maxlengthstayatclusters,minlengthstayatclusters,meanlengthstayatclusters,stdlengthstayatclusters,locationentropy,normalizedlocationentropy]
+                    ACCURACY_LIMIT:
+                      type: integer
+                      exclusiveMinimum: 0
+        additionalProperties:
```
Comment (JulioV) [inline tools/config.schema.yaml:565]: BARNETT location provider is missing

File: tools/config.schema.yaml [hunk 17]
```diff
@@ -0,0 +1,855 @@
...
+      TABLE: 
+        type: string
+      PROVIDERS:
+        type: ["null", object]
+        properties:
+          RAPIDS:
+            allOf:
+              - $ref: "#/definitions/PROVIDER"
+              - properties:
+                  REFERENCE_HOUR_FIRST_USE:
```
Comment (JulioV) [inline tools/config.schema.yaml:618]: 0-23 possible values

File: tools/config.schema.yaml [hunk 18]
```diff
@@ -0,0 +1,855 @@
...
+    required: [TABLE, PROVIDERS]
+    properties:
+      TABLE:
+        type: string
+      PROVIDERS:
+        type: ["null", object]
+        properties:
+          RAPIDS:
+            allOf:
+              - $ref: "#/definitions/PROVIDER"
```
Comment (JulioV) [inline tools/config.schema.yaml:669]: Missing the same feature validations as PHONE_WIFI_CONNECTED

File: tools/config.schema.yaml [hunk 19]
```diff
@@ -0,0 +1,855 @@
...
+          $ref: "#/definitions/PROVIDER"
+
+  FITBIT_DATA_CONFIGURATION:
+    allOf:
+      - $ref: "#/definitions/DATA_CONFIGURATION"
+      - properties:
+          SOURCE:
+            properties:
+              COLUMN_FORMAT: 
+                type: string
```
Comment (JulioV) [inline tools/config.schema.yaml:680]: This should be an enum JSON OR PLAIN_TEXT

File: tools/config.schema.yaml [hunk 20]
```diff
@@ -0,0 +1,855 @@
...
+
```
Comment (JulioV) [inline tools/config.schema.yaml:681]: Rebase develop in your branch so you can see the latest changes. There is a new sensor called FITBIT_DATA_YIELD that should be validated similarly to PHONE_DATA_YIELD

File: tools/config.schema.yaml [hunk 21]
```diff
@@ -0,0 +1,855 @@
...
+          $ref: "#/definitions/PROVIDER"
+
+  FITBIT_SLEEP_SUMMARY:
+    type: object
+    required: [TABLE, SLEEP_EPISODE_TIMESTAMP, PROVIDERS]
+    properties:
+      TABLE:
+        type: string
+      SLEEP_EPISODE_TIMESTAMP:
+        type: string
```
Comment (JulioV) [inline tools/config.schema.yaml:729]: enum: start or end

File: tools/config.schema.yaml [hunk 22]
```diff
@@ -0,0 +1,855 @@
...
+          enum: ["PHONE_ACCELEROMETER", "PHONE_ACTIVITY_RECOGNITION", "PHONE_APPLICATIONS_FOREGROUND", "PHONE_BATTERY", "PHONE_BLUETOOTH", "PHONE_CALLS", "PHONE_CONVERSATION", "PHONE_LIGHT", "PHONE_LOCATIONS", "PHONE_MESSAGES", "PHONE_SCREEN", "PHONE_WIFI_CONNECTED", "PHONE_WIFI_VISIBLE"]
+
+  HEATMAP_FEATURE_CORRELATION_MATRIX:
+    type: object
+    required: [PLOT, MIN_ROWS_RATIO, CORR_THRESHOLD, CORR_METHOD]
+    properties:
+      PLOT: 
+        type: boolean
+      MIN_ROWS_RATIOS:
+        type: number
```
Comment (JulioV) [inline tools/config.schema.yaml:851]: 0 to 1 number

File: tools/config.schema.yaml [hunk 23]
```diff
@@ -0,0 +1,855 @@
...
+      CORR_THRESHOLD: 
+        type: number
```
Comment (JulioV) [inline tools/config.schema.yaml:853]: 0 to 1 number

File: tools/config.schema.yaml [hunk 24]
```diff
@@ -0,0 +1,855 @@
...
+      CORR_METHOD:
+        type: string
```
Comment (JulioV) [inline tools/config.schema.yaml:855]: enum "pearson", "kendall", "spearman"

**Review by ChinW97 — APPROVED** [review]
Changes reviewed

File: tools/config.schema.yaml (hunk 24 above)
Comment (ChinW97) [reply tools/config.schema.yaml:855]: Added

File: tools/config.schema.yaml (hunk 23 above)
Comment (ChinW97) [reply tools/config.schema.yaml:853]: Added

File: tools/config.schema.yaml (hunk 22 above)
Comment (ChinW97) [reply tools/config.schema.yaml:851]: completed

File: tools/config.schema.yaml (hunk 21 above)
Comment (ChinW97) [reply tools/config.schema.yaml:729]: Added

File: tools/config.schema.yaml (hunk 6 above)
Comment (ChinW97) [reply tools/config.schema.yaml:271]: fixed

File: tools/config.schema.yaml (hunk 7 above)
Comment (ChinW97) [reply tools/config.schema.yaml:274]: fixed

File: tools/config.schema.yaml (hunk 9 above)
Comment (ChinW97) [reply tools/config.schema.yaml:298]: fixed

File: tools/config.schema.yaml (hunk 8 above)
Comment (ChinW97) [reply tools/config.schema.yaml:281]: fixed

File: tools/config.schema.yaml (hunk 10 above)
Comment (ChinW97) [reply tools/config.schema.yaml:338]: fixed

File: tools/config.schema.yaml (hunk 11 above)
Comment (ChinW97) [reply tools/config.schema.yaml:383]: Thanks!

File: tools/config.schema.yaml (hunk 12 above)
Comment (ChinW97) [reply tools/config.schema.yaml:469]: fixed

File: tools/config.schema.yaml (hunk 13 above)
Comment (ChinW97) [reply tools/config.schema.yaml:471]: fixed

File: tools/config.schema.yaml (hunk 14 above)
Comment (ChinW97) [reply tools/config.schema.yaml:482]: fixed

File: tools/config.schema.yaml (hunk 15 above)
Comment (ChinW97) [reply tools/config.schema.yaml:555]: fixed

File: tools/config.schema.yaml (hunk 16 above)
Comment (ChinW97) [reply tools/config.schema.yaml:565]: fixed

File: tools/config.schema.yaml (hunk 17 above)
Comment (ChinW97) [reply tools/config.schema.yaml:618]: fixed

File: tools/config.schema.yaml (hunk 18 above)
Comment (ChinW97) [reply tools/config.schema.yaml:669]: fixed

File: tools/config.schema.yaml (hunk 19 above)
Comment (ChinW97) [reply tools/config.schema.yaml:717]: fixed

File: tools/config.schema.yaml (hunk 1 above)
Comment (JulioV) [reply tools/config.schema.yaml:3]: Still missing most of the phone, fitbit and plot keys

File: tools/config.schema.yaml (hunk 5 above)
Comment (JulioV) [reply tools/config.schema.yaml:53]: this one is still missing

File: tools/config.schema.yaml (hunk 14 above)
Comment (JulioV) [reply tools/config.schema.yaml:481]: the new sensors are missing like PHONE_APPLICATIONS_CRASHES

File: tools/config.schema.yaml [hunk 25]
```diff
@@ -0,0 +1,923 @@
...
+              - properties:
+                  SINGLE_CATEGORIES: 
+                    type: array
+                    uniqueItems: True
+                    items:
+                      type: string
+                  MULTIPLE_CATEGORIES:
+                    type: object
+                    properties:
+                      social:
```
Comment (JulioV) [inline tools/config.schema.yaml:274]: we only need to validate that `properties` of `MULTIPLE_CATEGORIES`  are string arrays but without specifying any array like `social`

File: tools/config.schema.yaml [hunk 26]
```diff
@@ -0,0 +1,923 @@
...
+                    FEATURES: 
+                      type: array
+                      uniqueItems: True
+                      items:
+                        type: string
+                        enum: [locationvariance,loglocationvariance,totaldistance,averagespeed,varspeed,circadianmovement,numberofsignificantplaces,numberlocationtransitions,radiusgyration,timeattop1location,timeattop2location,timeattop3location,movingtostaticratio,outlierstimepercent,maxlengthstayatclusters,minlengthstayatclusters,meanlengthstayatclusters,stdlengthstayatclusters,locationentropy,normalizedlocationentropy]
+                    ACCURACY_LIMIT:
+                      type: integer
+                      exclusiveMinimum: 0
+                    DBSCAN_EPS:
```
Comment (JulioV) [inline tools/config.schema.yaml:565]: >0

File: tools/config.schema.yaml [hunk 27]
```diff
@@ -0,0 +1,923 @@
...
+                      type: integer
+                    DBSCAN_MINSAMPLES:
```
Comment (JulioV) [inline tools/config.schema.yaml:567]: >0

File: tools/config.schema.yaml [hunk 28]
```diff
@@ -0,0 +1,923 @@
...
+                      type: integer
+                    THRESHOLD_STATIC:
```
Comment (JulioV) [inline tools/config.schema.yaml:569]: >0

File: tools/config.schema.yaml [hunk 29]
```diff
@@ -0,0 +1,923 @@
...
+                      type: integer
+                    MAXIMUM_GAP_ALLOWED:
```
Comment (JulioV) [inline tools/config.schema.yaml:571]: >0

File: tools/config.schema.yaml [hunk 30]
```diff
@@ -0,0 +1,923 @@
...
+                      type: integer
+                    MINUTES_DATA_USED:
+                      type: boolean
+                    SAMPLING_FREQUENCY:
+                      type: integer
+                    CLUSTER_ON:
```
Comment (JulioV) [inline tools/config.schema.yaml:577]: enum

File: tools/config.schema.yaml [hunk 31]
```diff
@@ -0,0 +1,923 @@
...
+                      type: string
+                    CLUSTERING_ALGORITHM:
```
Comment (JulioV) [inline tools/config.schema.yaml:579]: enum

File: tools/config.schema.yaml (hunk 18 above)
Comment (JulioV) [reply tools/config.schema.yaml:706]: still missing the feature validations

File: tools/config.schema.yaml [hunk 32]
```diff
@@ -0,0 +1,923 @@
...
+              COLUMN_FORMAT: 
+                  type: string
+                  enum: ["JSON", "PLAIN_TEXT"]
+  
+  FITBIT_DATA_YIELD:
+    type: object
+    required: [SENSOR, PROVIDERS]
+    properties:
+      SENSOR:
+        type: string
```
Comment (JulioV) [inline tools/config.schema.yaml:725]: enum, only possible value atm is `FITBIT_HEARTRATE_INTRADAY`

**Review by JulioV — COMMENTED** [review]
few more changes

File: tools/config.schema.yaml (hunk 5 above)
Comment (ChinW97) [reply tools/config.schema.yaml:53]: Fixed

File: tools/config.schema.yaml (hunk 25 above)
Comment (ChinW97) [reply tools/config.schema.yaml:274]: Fixed

File: tools/config.schema.yaml (hunk 26 above)
Comment (ChinW97) [reply tools/config.schema.yaml:565]: Fixed

File: tools/config.schema.yaml (hunk 27 above)
Comment (ChinW97) [reply tools/config.schema.yaml:567]: fixed

File: tools/config.schema.yaml (hunk 28 above)
Comment (ChinW97) [reply tools/config.schema.yaml:569]: Fixed

File: tools/config.schema.yaml (hunk 29 above)
Comment (ChinW97) [reply tools/config.schema.yaml:571]: Fixed

File: tools/config.schema.yaml (hunk 30 above)
Comment (ChinW97) [reply tools/config.schema.yaml:577]: fixed

File: tools/config.schema.yaml (hunk 31 above)
Comment (ChinW97) [reply tools/config.schema.yaml:579]: fixed

File: tools/config.schema.yaml (hunk 32 above)
Comment (ChinW97) [reply tools/config.schema.yaml:725]: fixed

File: tools/config.schema.yaml (hunk 1 above)
Comment (ChinW97) [reply tools/config.schema.yaml:3]: Missing time_segment, will add it later

Comment (ChinW97) [reply tools/config.schema.yaml:3]: Fixed

**Comment by JulioV** [comment]
Merged in bef868671ef9cde4d78dcde1039272e0cae09f1e

**Pull request closed without merging**